lib/mapObjects/CGHeroInstance.cpp

```diff
@@ -129,14 +129,9 @@ int3 CGHeroInstance::getPosition(bool h3m) const //h3m=true - returns position o
 	}
 }
 
-bool CGHeroInstance::canFly() const
+const Bonus * CGHeroInstance::getBonusAtTurn(const Bonus::BonusType &type, const int &turn, const TBonusSubtype &subType) const
 {
-	return hasBonusOfType(Bonus::FLYING_MOVEMENT);
-}
-
-bool CGHeroInstance::canWalkOnSea() const
-{
-	return hasBonusOfType(Bonus::WATER_WALKING);
+	return getBonus(Selector::type(type).And(Selector::days(turn)).And(Selector::subtype(subType)));
```

Comment:
Didn't yet find out with bonus system so not sure if "cachingStr" intended to be passed outside of bonus system. In same time that my function is too limited to pathfinder to pathfinding to put it inside HeroBonus.

Though I see you using cachingStr in CSpell::internalIsImmune.